python/tank/util/metrics.py

```diff
@@ -324,100 +327,196 @@ def _dispatch(self, metrics):
+            # assigning the dictionnary so we can pre-inpect and sanitize
+            # if necessary.
+            for key in properties.keys():
+                value = properties[key]
+                self._add_event_property(key, value)
 
-    def __repr__(self):
-        """Official str representation of the user activity metric."""
-        return "%s.%s" % (self._data["module"], self._data["action"])
 
+        self._add_system_info_properties()
 
-class UserAttributeMetric(ToolkitMetric):
-    """Convenience class for a user attribute metric."""
+    def _add_event_property(self, name, value):
+        """
+        Helper method allowing some verification and sanitization of the
+        specified properties.
 
-    def __init__(self, attr_name, attr_value):
-        """Initialize the metric with the attribute name and value.
-        
-        :param str attr_name: Name of the attribute.
-        :param str attr_value: The value of the attribute.
+        :param str name:
+        :param object value:
+        """
+        # TODO: escape non-ascii7 strings here or in Dispatcher ?
+        self._data["event_property"][name] = value
 
+    def _add_system_info_properties(self):
+        """
+        Helper method adding a number of system information metrics such as:
+        - OS type
+        - OS version
+        - System architechture
+        - Manufacturer etc.
+
+        The method takes care of gathering relevant information and properly
+        formatting them for submission to a Shotgun site
         """
-        super(UserAttributeMetric, self).__init__({
-            "type": "user_attribute",
-            "attr_name": attr_name,
-            "attr_value": attr_value,
-        })
+
+        # TODO: find a descent library and implement method
+        # TODO: Can't really send Amplitude-usable system information metrics
+        #       as we can only send 'event_properties'
+        pass
 
     def __repr__(self):
-        """Official str representation of the user attribute metric."""
-        return "%s.%s" % (self._data["attr_name"], self._data["attr_value"])
+        """Official str representation of the user activity metric."""
+        return "%s:%s" % (self._data["event_group"], self._data["event_name"])
+
+    def __str__(self):
+        """Readable str representation of the metric."""
+        return "%s: %s" % (self.__class__, self._data)
+
+    @property
+    def data(self):
+        """The underlying data this metric represents."""
+        return self._data
 
 
 ###############################################################################
 # metrics logging convenience functions
 
-def log_metric(metric, log_once=False):
-    """Log a Toolkit metric.
-    
-    :param ToolkitMetric metric: The metric to log.
-    :param bool log_once: ``True`` if this metric should be ignored if it has
-        already been logged. Defaults to ``False``.
+def log_event_metric(metric_event, log_once=False):
```

Comment:
I think I like the `@classmethod` better. It'll simplify the imports. I think the `GROUP_*` and `KEY_*` attributes justify having the class for namespacing. I'm ok with having the object creation happen under the hood as well. Having this be a separate method doesn't really buy us much I think.